compiler/damlc/daml-lf-conversion/src/DA/Daml/LFConversion.hs

```diff
@@ -1523,6 +1531,162 @@ toCtor env con =
 isRecordCtor :: Ctor -> Bool
 isRecordCtor (Ctor _ fldNames fldTys) = not (null fldNames) || null fldTys
 
+------------------------------------------------------------------------------
+-- EXTERNAL PACKAGES
+
+-- External instance methods
+convertExternal :: Env -> LF.PackageRef -> String -> LF.Type -> LF.Expr
+convertExternal env stdlibRef primId lfType
+    | [pkgId, modStr, templName, method] <- splitOn ":" primId
+    , Just LF.Template {..} <- lookup pkgId modStr templName =
+        let pkgRef = PRImport $ PackageId $ T.pack pkgId
+            mod = ModuleName $ map T.pack $ splitOn "." modStr
+            qualify tconName =
+                Qualified
+                    { qualPackage = pkgRef
+                    , qualModule = mod
+                    , qualObject = tconName
+                    }
+            templateDataType = TCon . qualify
+         in case method of
+                "signatory" ->
+                    ETmLam
+                        (tplParam, templateDataType tplTypeCon)
+                        tplSignatories
+                "observer" ->
+                    ETmLam (tplParam, templateDataType tplTypeCon) tplObservers
+                "agreement" ->
+                    ETmLam (tplParam, templateDataType tplTypeCon) tplAgreement
+                "ensure" ->
+                    ETmLam
+                        (tplParam, templateDataType tplTypeCon)
+                        tplPrecondition
+                "create" ->
+                    ETmLam
+                        (tplParam, templateDataType tplTypeCon)
+                        (EUpdate $ UCreate (qualify tplTypeCon) (EVar tplParam))
+                "fetch" ->
+                    let coid = mkVar "$coid"
+                     in ETmLam
+                            ( coid
+                            , TApp
+                                  (TBuiltin BTContractId)
+                                  (templateDataType tplTypeCon))
+                            (EUpdate $ UFetch (qualify tplTypeCon) (EVar coid))
+                "archive" ->
+                    let archiveChoice = ChoiceName "Archive"
+                     in case NM.lookup archiveChoice tplChoices of
+                            Nothing -> errorExpr "archive"
+                            Just TemplateChoice {..} ->
+                                case chcArgBinder of
```

Comment:
This code is somewhat sketchy. I can give you a DALF where `archive` does something completely different and you will happily generate a `Template` instance using this choice as the `archive` implementation. I don’t have a great solution here apart from verifying that this choice is actually the archive choice that we expect. Another hint that archive should really be a primitive.

Reply:
I think that would actually be fine. We would archive the template in the way the signatories agreed to upfront. But I wouldn't mind if it would be a primitive either so that a weird `archive` can't happen.

Reply:
It’s fine in the sense that you can’t do anything people haven’t agreed to before but it means that `archive` is now no longer guaranteed to archive a template, e.g., I could give you a nonconsuming choice that doesn’t archive anything.